linkd/src/args.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+// TODO(xla): Expose discovery args.
+// TODO(xla): Expose protocol args (membership, network, replication, etc.).
+// TODO(xla): Expose storage args.
+// TODO(xla): Expose logging args.
+
+use std::{fmt, path::PathBuf, str::FromStr};
+
+use structopt::StructOpt;
+
+use librad::{
+    crypto,
+    profile::{ProfileId, RadHome},
+    PeerId,
+};
+
+#[derive(Debug, Default, Eq, PartialEq, StructOpt)]
+pub struct Args {
+    /// List of bootstrap nodes for initial discovery.
+    #[structopt(long = "bootstrap", name = "bootstrap")]
+    pub bootstraps: Vec<Bootstrap>,
+
+    /// Identifier of the profile the daemon will run for. This value determines
+    /// which monorepo (if existing) on disk will be the backing storage.
+    #[structopt(long)]
+    pub profile_id: Option<ProfileId>,
+
+    /// Home of the profile data, if not provided is read from the environment
+    /// and falls back to project dirs.
+    #[structopt(long, default_value, parse(from_str = parse_rad_home))]
+    pub rad_home: RadHome,
```

Comment:
This takes precedence.

Reply:
But if it's not specified then it defaults to `RadHome::ProjectDirs`, so I could do:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```
and then it would end up using the project dirs

Reply:
We could employ structopts magic of reading env variables. I wonder tho, will this be hard to communicate what arg will have what value - the matrix of combinations is expanding. Furthermore the daemon is unlikely be run just from a users tty. Where and when is it usual to set the `RAD_HOME`?

Reply:
> Where and when is it usual to set the RAD_HOME?

I would say that it should be used for testing or someone who _really_ wants their home to be somewhere else. So I think this parameter should just be removed. 

For its use here[[0]], I would recommend this usage[[1]].

[0]: https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131
[1]: https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72

Reply:
I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mi of env vars and cli options.

Reply:
> For its use here[[0](https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131)], I would recommend this usage[[1](https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72)].

Sorry I was looking for places where the env var `RAD_HOME` is set.

Reply:
> I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mix of env vars and cli options.

I think in this particular case it's not going to be set that often. So you won't realistically be mixing env and cli options. 

> Sorry I was looking for places where the env var RAD_HOME is set.

Upstream's sandbox environment :shrug: It doesn't tend to be set for actual use.

Reply:
> Upstream's sandbox environment shrug It doesn't tend to be set for actual use.

Reinforces my impression to try to weave in these env vars and always expect explicit configuration through the cli.

Reply:
Then you need to find a way for this CLI parameter to play well with the env var

Reply:
> Then you need to find a way for this CLI parameter to play well with the env var

If it's never set, why bother keeping it the env var around?

Reply:
It doesn't mean it _can't_ be set. Like I said above, the following command would have undesirable effects:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```

Reply:
1b60712